akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala

```diff
@@ -0,0 +1,685 @@
+      current = traversalStack.removeLast()
+
+      while (current ne EmptyTraversal) {
+        var nextStep: Traversal = EmptyTraversal
+        current match {
+          case MaterializeAtomic(mod, outToSlot) ⇒
+            if (Debug) println(s"materializing module: $mod")
+            val matAndStage = islandTracking.getCurrentPhase.materializeAtomic(mod, attributesStack.getLast)
+            val logic = matAndStage._1
+            val matValue = matAndStage._2
+            if (Debug) println(s"  materialized value is $matValue")
+            matValueStack.addLast(matValue)
+
+            val ins = mod.shape.inlets.iterator
+            val stageGlobalOffset = islandTracking.getCurrentOffset
+
+            while (ins.hasNext) {
+              val in = ins.next()
+              islandTracking.wireIn(in, logic)
+            }
+
+            val outs = mod.shape.outlets.iterator
+            while (outs.hasNext) {
+              val out = outs.next()
+              val absoluteTargetSlot = stageGlobalOffset + outToSlot(out.id)
+              if (Debug) println(s"  wiring offset: ${outToSlot.mkString("[", ",", "]")}")
+              islandTracking.wireOut(out, absoluteTargetSlot, logic)
+            }
+
+            if (Debug) println(s"PUSH: $matValue => $matValueStack")
+
+          case Concat(first, next) ⇒
+            if (next ne EmptyTraversal) traversalStack.add(next)
+            nextStep = first
+          case Pop ⇒
+            val popped = matValueStack.removeLast()
+            if (Debug) println(s"POP: $popped => $matValueStack")
+          case PushNotUsed ⇒
+            matValueStack.addLast(NotUsed)
+            if (Debug) println(s"PUSH: NotUsed => $matValueStack")
+          case Transform(f) ⇒
+            val prev = matValueStack.removeLast()
+            val result = f(prev)
+            matValueStack.addLast(result)
+            if (Debug) println(s"TRFM: $matValueStack")
+          case Compose(f) ⇒
+            val second = matValueStack.removeLast()
+            val first = matValueStack.removeLast()
+            val result = f(first, second)
+            matValueStack.addLast(result)
+            if (Debug) println(s"COMP: $matValueStack")
+          case PushAttributes(attr) ⇒
+            attributesStack.addLast(attributesStack.getLast and attr)
+            if (Debug) println(s"ATTR PUSH: $attr")
+          case PopAttributes ⇒
+            attributesStack.removeLast()
+            if (Debug) println(s"ATTR POP")
+          case EnterIsland(tag, island) ⇒
+            traversalStack.addLast(islandTracking.enterIsland(tag))
+            nextStep = island
+          case ex: ExitIsland ⇒
+            islandTracking.exitIsland(ex)
+          case _ ⇒
+        }
+        current = nextStep
+      }
+    }
+
+    islandTracking.getCurrentPhase.onIslandReady()
+
+    if (Debug) println("--- Finished materialization")
+    matValueStack.peekLast().asInstanceOf[Mat]
+  }
+
+}
+
+trait IslandTag
```

Comment:
No, this will be public API so people can plug in their own phases. I create a ticket for properly finalizing this API (can be post 2.5.0).